src/Datadog.Trace.DuckTyping/DuckType.Fields.cs

```diff
@@ -0,0 +1,189 @@
+                        dynIL.Emit(targetField.DeclaringType!.IsValueType ? OpCodes.Unbox : OpCodes.Castclass, targetField.DeclaringType);
+                    }
+                }
+
+                // Emit the field and convert before returning (in case of boxing)
+                dynIL.Emit(targetField.IsStatic ? OpCodes.Ldsfld : OpCodes.Ldfld, targetField);
+                ILHelpers.TypeConversion(dynIL, targetField.FieldType, returnType);
+                dynIL.Emit(OpCodes.Ret);
+
+                // Emit the Call to the dynamic method pointer [Calli]
+                il.Emit(OpCodes.Ldc_I8, (long)GetRuntimeHandle(dynMethod).GetFunctionPointer());
+                il.Emit(OpCodes.Conv_I);
+                il.EmitCalli(OpCodes.Calli, dynMethod.CallingConvention, returnType, dynParameters, null);
+            }
+
+            // Check if the type can be converted or if we need to enable duck chaining
+            if (proxyProperty.PropertyType != targetField.FieldType && !proxyProperty.PropertyType.IsValueType && !proxyProperty.PropertyType.IsAssignableFrom(targetField.FieldType))
+            {
+                // We call DuckType.CreateCache<>.Create()
+                MethodInfo getProxyMethodInfo = typeof(CreateCache<>)
+                    .MakeGenericType(proxyProperty.PropertyType).GetMethod("Create");
+
+                il.Emit(OpCodes.Call, getProxyMethodInfo);
+            }
+            else if (returnType != proxyProperty.PropertyType)
+            {
+                // If the type is not the expected type we try a conversion.
+                ILHelpers.TypeConversion(il, returnType, proxyProperty.PropertyType);
+            }
+
+            il.Emit(OpCodes.Ret);
+            return proxyMethod;
+        }
+
+        private static MethodBuilder GetFieldSetMethod(TypeBuilder proxyTypeBuilder, Type targetType, PropertyInfo proxyProperty, FieldInfo targetField, FieldInfo instanceField)
+        {
+            MethodBuilder method = proxyTypeBuilder.DefineMethod(
+                "set_" + proxyProperty.Name,
+                MethodAttributes.Public | MethodAttributes.SpecialName | MethodAttributes.Final | MethodAttributes.HideBySig | MethodAttributes.Virtual,
+                typeof(void),
+                new[] { proxyProperty.PropertyType });
+
+            ILGenerator il = method.GetILGenerator();
+            bool isPublicInstance = targetType.IsPublic || targetType.IsNestedPublic;
+            Type currentValueType = proxyProperty.PropertyType;
+
+            // Load instance
+            if (!targetField.IsStatic)
+            {
+                il.Emit(OpCodes.Ldarg_0);
+                il.Emit(OpCodes.Ldfld, instanceField);
+            }
+
+            // Check if the type can be converted of if we need to enable duck chaining
+            if (proxyProperty.PropertyType != targetField.FieldType && !proxyProperty.PropertyType.IsValueType && !proxyProperty.PropertyType.IsAssignableFrom(targetField.FieldType))
```

Comment:
In both branches of this if/else, if the method is static, then this dynamic method will only have one argument (index 0), so it looks like `il.Emit(OpCodes.Ldarg_1)` is incorrect. Or am I missing something?

Reply:
In this case we are not emitting inside a `DynamicMethod` but inside the `MethodBuilder` for the `ProxyType` being generated, so because we are actually implementing an interface (the proxy definition) we will always have an instance, the proxy instance as `OpCodes.Ldarg_0` and the value to passthrough to the field as `OpCodes.Ldarg_1`.

When we emit a `DynamicMethod` to bypass the visibility checks in the same file, we actually do what you're describing:  https://github.com/DataDog/dd-trace-dotnet/blob/87582c6a5a67c4377a7fc59b017a8ec5eec303da/src/Datadog.Trace.DuckTyping/DuckType.Fields.cs#L157-L177

Emit `OpCodes.Ldarg_0` and depending if is an static use that value from the stack, if not, we load `OpCodes.Ldarg_1`.

Reply:
Oh I see, this makes sense. Thanks for explaining that!